src/mpid/ch4/netmod/ofi/ofi_am_events.h

```diff
@@ -103,6 +103,51 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_handle_short_am(MPIDI_OFI_am_header_t * m
     return mpi_errno;
 }
 
+/* this is called in am_recv_event in ofi_event.c on receiver side */
+MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_handle_pipeline(MPIDI_OFI_am_header_t * msg_hdr,
+                                                       void *am_hdr, void *p_data)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int is_done = 0;
+    MPIR_Request *rreq = NULL;
+    MPIR_Request *cache_rreq = NULL;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_HANDLE_PIPELINE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_HANDLE_PIPELINE);
+
+    cache_rreq = MPIDIG_req_cache_lookup(MPIDI_OFI_global.req_map, (uint64_t) msg_hdr->fi_src_addr);
+    MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                    (MPL_DBG_FDEST, "cached req %p handle=0x%x", cache_rreq,
+                     cache_rreq ? cache_rreq->handle : 0));
+
+    rreq = cache_rreq;
+
+    if (!rreq) {
+        /* no cached request, this must be the first segment */
+        /* note: setting is_local, is_async, req to 0, 1, rreq */
+        MPIDIG_global.target_msg_cbs[msg_hdr->handler_id] (msg_hdr->handler_id, am_hdr, p_data,
+                                                           msg_hdr->data_sz, 0, 1, &rreq);
+        MPIDIG_recv_setup(rreq);
+    }
+
+    is_done = MPIDIG_recv_copy_seg(p_data, msg_hdr->seg_sz, rreq);
+    if (is_done) {
+        MPIDIG_REQUEST(rreq, req->target_cmpl_cb) (rreq);
+        MPIDIG_req_cache_remove(MPIDI_OFI_global.req_map, (uint64_t) msg_hdr->fi_src_addr);
+    } else {
+        /* only cache the request if we do expect more data. This is a corner case when
+         * data+send_hdr is greater than eager limit but data+send_data_hdr is smaller than eager
+         * limit which results in single segment of the pipeline. */
+        if (!cache_rreq) {
+            MPIDIG_req_cache_add(MPIDI_OFI_global.req_map, (uint64_t) msg_hdr->fi_src_addr, rreq);
+        }
```

Comment:
Intuitively, one expect to always add cache on first segment. For the described "corner" case, it just mean we'll add cache then remove cache immediately. Nest the adding cache inside the branch here makes the code a bit more obscure. Pipeline is not the performance critical path, so I would recommend opt for clarity.

Reply:
alright